File: backend/packages/python/flow_map/exceptions.py
class FlowMapInitializationError(RuntimeError):
    def __init__(self, message: str) -> None:
        super().__init__(f"[FlowMap] Initialization failed: {message}")


class FlowMapConfigError(ValueError):
    def __init__(self, field: str, reason: str) -> None:
        super().__init__(f'[FlowMap] Invalid configuration for "{field}": {reason}')


class AstParserError(RuntimeError):
    def __init__(self, file_path: str, reason: str) -> None:
        super().__init__(f'[AstParser] Failed to parse "{file_path}": {reason}')


class NanoAgentError(RuntimeError):
    def __init__(self, node_id: str, reason: str) -> None:
        super().__init__(f'[NanoAgent] Failed to summarize node "{node_id}": {reason}')


class CacheError(RuntimeError):
    def __init__(self, operation: str, reason: str) -> None:
        super().__init__(f"[Cache] Failed to {operation}: {reason}")


class SidecarError(RuntimeError):
    def __init__(self, port: int, reason: str) -> None:
        super().__init__(f"[Sidecar] Failed to start on port {port}: {reason}")

File: backend/packages/nextjs/src/exceptions/exceptions.ts
export class CodeMapInitializationException extends Error {
  constructor(message: string) {
    super(`[code.map] Initialization failed: ${message}`);
    this.name = 'CodeMapInitializationException';
  }
}

export class CodeMapConfigException extends Error {
  constructor(field: string, reason: string) {
    super(`[code.map] Invalid configuration for "${field}": ${reason}`);
    this.name = 'CodeMapConfigException';
  }
}

export class AstParserException extends Error {
  constructor(filePath: string, reason: string) {
    super(`[AstParser] Failed to parse "${filePath}": ${reason}`);
    this.name = 'AstParserException';
  }
}

export class NanoAgentException extends Error {
  constructor(nodeId: string, reason: string) {
    super(`[NanoAgent] Failed to summarize node "${nodeId}": ${reason}`);
    this.name = 'NanoAgentException';
  }
}

export class CacheException extends Error {
  constructor(operation: string, reason: string) {
    super(`[Cache] Failed to ${operation}: ${reason}`);
    this.name = 'CacheException';
  }
}

export class SidecarException extends Error {
  constructor(port: number, reason: string) {
    super(`[Sidecar] Failed to start on port ${port}: ${reason}`);
    this.name = 'SidecarException';
  }
}
